tensorly/tenalg/proximal.py

```diff
@@ -91,4 +91,316 @@ def procrustes(matrix):
+    cnt: integer
+        final number of update iteration performed
+    rho: float
+        number of loops authorized by the time stop criterion
+
+    References
+    ----------
+    [1]: N. Gillis and F. Glineur, Accelerated Multiplicative Updates and
+    Hierarchical ALS Algorithms for Nonnegative Matrix Factorization,
+    Neural Computation 24 (4): 1085-1105, 2012.
+
+    [2] J. Eggert, and E. Korner. "Sparse coding and NMF."
+    2004 IEEE International Joint Conference on Neural Networks
+    (IEEE Cat. No. 04CH37541). Vol. 4. IEEE, 2004.
+
+    """
+
+    r, n = tl.shape(UtM)
+    if in_V is None:  # checks if V is empty
+        V = tl.solve(UtU, UtM)
+
+        V[V < 0] = 0
+        # Scaling
+        scale = tl.sum(UtM * V) / tl.sum(
+            UtU * tl.dot(V, tl.transpose(V)))
+        V = tl.dot(scale, V)
+    else:
+        V = in_V
+
+    rho = 1
+    eps0 = 0
+    cnt = 1
+    eps = 1
+
+
+    while eps >= delta * eps0 and cnt <= 1 + 0.5* rho and cnt <= maxiter:
+        nodelta = 0
+        for k in range(r):
+
+            if UtU[k, k] != 0:
+                if sparsity_coefficient != None: # Modifying the objective function for sparsification
+                    if tl.get_backend() == 'pytorch':
+                        import torch
+                        deltaV = torch.maximum((UtM[k, :] - UtU[k, :] @ V - sparsity_coefficient * tl.ones(n)) / UtU[k, k],
+                                        -V[k, :])
+                    else:
+                        deltaV = tl.max([(UtM[k, :] - UtU[k, :] @ V - sparsity_coefficient * tl.ones(n)) / UtU[k, k],
+                                        -V[k, :]],axis=0)
+                    if tl.get_backend()=='tensorflow':
+                        import tensorflow as tf
+                        V=tf.Variable(V,dtype='float')
+                        V[k, :].assign(V[k, :] + deltaV)
+                    else:
+                        V[k, :] = V[k, :] + deltaV
+
+                else:  # without sparsity
+
+                    if tl.get_backend() == 'pytorch':
+                        import torch
+                        deltaV = torch.maximum((UtM[k, :] - tl.dot(UtU[k, :], V)) / UtU[k, k],
+                                         -V[k, :])
+                    else:
+                        deltaV = tl.max([(UtM[k, :] - tl.dot(UtU[k, :], V)) / UtU[k, k],
+                                         -V[k, :]], axis=0)
+                    if tl.get_backend()=='tensorflow':
+                        import tensorflow as tf
+                        V=tf.Variable(V,dtype='float')
+                        V[k, :].assign(V[k, :] + deltaV)
+                    else:
+                        V[k, :] = V[k, :] + deltaV
+
+                nodelta = nodelta + tl.dot(deltaV, tl.transpose(deltaV))
+
+                # Safety procedure, if columns aren't allow to be zero
+                if nonzero_row and (V[k, :] == 0).all():
+                    V[k, :] = 1e-16 * tl.max(V)
```

Comment:
You could use `tl.eps(V.dtype)`, or check how we checked for zero values in cp_normalize.

Reply:
As I understand, you mean `(V[k, :] == 0).all()` part. Now, I changed it to  `tl.all(V[k, :] == 0)`.

Reply:
I also was talking about the 1e-16: you could use `tl.eps(V.dtype)` instead of that hard-coded value so it matches the dtype of V.

Reply:
ok, I changed that part as well